### PR TITLE
Feature advanced undo redo

### PR DESCRIPTION
**Pull request by TAEUK-PARK**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -3,6 +3,7 @@
   "version": "0.1.0",
   "private": true,
   "dependencies": {
+    "@codemirror/history": "^0.19.2",
```
Comment (yesohyeon) [inline package.json]: 필요한가요?

File: src/components/CodeArea.js [hunk 2]
```diff
@@ -34,19 +34,32 @@ export default function CodeArea({
...
     html: { content: htmlCode },
     css: { content: cssCode },
     js: { content: jsCode },
   } = wholeCode;
 
-  const saveContent = (text, prevText, language) => {
+  const [isTrans, setIsTrans] = useState({
+    html: false,
+    css: false,
+    js: false,
```
Comment (yesohyeon) [inline src/components/CodeArea.js]: isTransactionState

File: package.json (hunk 1 above)
Comment (yesohyeon) [inline package.json]: 필요한가요?

File: src/components/CodeArea.js [hunk 3]
```diff
@@ -57,16 +70,29 @@ export default function CodeArea({
         [language]: currentContent,
       };
     });
+    setIsTrans((prevState) => {
+      return { ...prevState, [selectedLanguage]: false };
+    });
```
Comment (yesohyeon) [inline src/components/CodeArea.js]: inline

File: src/components/CodeArea.js [hunk 4]
```diff
@@ -94,9 +132,48 @@ export default function CodeArea({
...
+      const { anchor, head } = selection[selectedLanguage];
+
+      handleChange((prevState) => {
+        prevState[selectedLanguage].anchor = anchor;
+        prevState[selectedLanguage].head = head;
+        return prevState;
+      });
+      setIsTrans((prevState) => {
+        return { ...prevState, [selectedLanguage]: true };
+      });
```
Comment (yesohyeon) [inline src/components/CodeArea.js]: inline

File: src/components/CodeArea.js [hunk 5]
```diff
@@ -94,9 +132,48 @@ export default function CodeArea({
...
+    }
+
+    setCurrentCode((prevState) => {
+      return { ...prevState, [selectedLanguage]: value };
+    });
```
Comment (yesohyeon) [inline src/components/CodeArea.js]: inline

File: src/components/FunctionHeader/RedoButton/RedoButton.js [hunk 6]
```diff
@@ -3,15 +3,31 @@ import PropTypes from "prop-types";
...
     });
+
+    const { anchor, head } = code[selectedLanguage].next;
+
+    handleUndoRedoClick((prevState) => {
+      return {
+        ...prevState,
+        [selectedLanguage]: { anchor, head },
+      };
+    });
```
Comment (yesohyeon) [inline src/components/FunctionHeader/RedoButton/RedoButton.js]: inline

File: src/components/FunctionHeader/UndoButton/UndoButton.js [hunk 7]
```diff
@@ -3,15 +3,29 @@ import PropTypes from "prop-types";
...
   const handleUndoClick = () => {
+    view.focus();
+
     if (!code[selectedLanguage].prev) return;
     handleClick((prevState) => {
       return {
         ...prevState,
         [selectedLanguage]: prevState[selectedLanguage].prev,
       };
     });
```
Comment (yesohyeon) [inline src/components/FunctionHeader/UndoButton/UndoButton.js]: inline

File: src/components/FunctionHeader/UndoButton/UndoButton.js [hunk 8]
```diff
@@ -28,4 +42,6 @@ UndoButton.propTypes = {
   code: PropTypes.object.isRequired,
   handleClick: PropTypes.func.isRequired,
   selectedLanguage: PropTypes.string.isRequired,
+  handleUndoRedoClick: PropTypes.func,
+  view: PropTypes.object,
```
Comment (yesohyeon) [inline src/components/FunctionHeader/UndoButton/UndoButton.js]: isRequired?

File: src/components/FunctionHeader/FunctionHeader.js [hunk 9]
```diff
@@ -35,6 +41,8 @@ FunctionHeader.propTypes = {
   handleClick: PropTypes.func.isRequired,
   selectedLanguage: PropTypes.string.isRequired,
   handleRunClick: PropTypes.func.isRequired,
+  handleUndoRedoClick: PropTypes.func,
+  view: PropTypes.object,
```
Comment (yesohyeon) [inline src/components/FunctionHeader/FunctionHeader.js]: isRequired?

File: src/components/FunctionHeader/RedoButton/RedoButton.js [hunk 10]
```diff
@@ -28,4 +44,6 @@ RedoButton.propTypes = {
   code: PropTypes.object.isRequired,
   handleClick: PropTypes.func.isRequired,
   selectedLanguage: PropTypes.string.isRequired,
+  handleUndoRedoClick: PropTypes.func,
+  view: PropTypes.object,
```
Comment (yesohyeon) [inline src/components/FunctionHeader/RedoButton/RedoButton.js]: isRequired?

File: package.json (hunk 1 above)
Comment (developerjessie) [inline package.json]: 사용하지 않는 dependency는 uninstall해주세요

**Review by yesohyeon — APPROVED** [review]
LGTM

**Pull request merged**